src/Raven.Server/Documents/Queries/WhereExpressionVisitor.cs

```
@@ -17,69 +19,99 @@ protected WhereExpressionVisitor(string queryText)
-                    VisitFieldToken(QueryExpression.Extract(expression.Field), expression.Value ?? expression.First, parameters);
-                    return;
-                case OperatorType.Between:
-                    VisitFieldTokens(QueryExpression.Extract(expression.Field), expression.First, expression.Second, parameters);
-                    return;
-                case OperatorType.In:
-                case OperatorType.AllIn:
-                    VisitFieldTokens(QueryExpression.Extract(expression.Field), expression.Values, parameters);
-                    return;
-                case OperatorType.Method:
-                    VisitMethodTokens(expression, parameters);
+                    VisitFieldToken(be.Left, be.Right, parameters);
                     return;
+                case OperatorType.And:
+                case OperatorType.AndNot:
+                case OperatorType.Or:
+                case OperatorType.OrNot:
+                    Visit(be.Left, parameters);
+                    Visit(be.Right, parameters);
+                    break;
+
                 default:
-                    throw new ArgumentException(expression.Type.ToString());
+                    ThrowInvalidOperatorType(expression);
+                    break;
             }
         }
 
-        protected ValueTokenType GetValueTokenType(BlittableJsonReaderObject parameters, ValueToken value, bool unwrapArrays)
+        private static void ThrowInvalidOperatorType(QueryExpression expression)
         {
-            var valueType = value.Type;
+            throw new ArgumentException(expression.Type.ToString());
+        }
 
-            if (valueType == ValueTokenType.Parameter)
-            {
-                var parameterName = QueryExpression.Extract(value);
+        private static void ThrowUnexpectedExpression(QueryExpression expression)
+        {
+            throw new InvalidOperationException("Expected binary expression, but got " + expression);
```

Maybe better to throw `InvalidQueryException` here?

---

Yes